### PR TITLE
Add StrokeLineWithColorAndThickness message to RemoteDisplayListRecorder

### PR DESCRIPTION
**Pull request by mattwoodrow**
#### 0eab85d1cfbec52489bb70f7bf2ad507c27d0fd1
<pre>
Add StrokeLineWithColorAndThickness message to RemoteDisplayListRecorder
<a href="https://bugs.webkit.org/show_bug.cgi?id=242772">https://bugs.webkit.org/show_bug.cgi?id=242772</a>

Reviewed by Simon Fraser.

There&apos;s some overhead per-message sent, and stroking lines is frequently combined with
updating the current stroke color and line thickness. Adding a special combination message
for this common case helps improve performance.

This also reduces the batch size, since we&apos;re now sending fewer (but larger) messages, and
having the batch size be too large results in unnecessary idle time on the receiving end.

This solution won&apos;t scale well, but hopefully is sufficient to handle the most common cases
in the short term.

I think an ideal solution would be to dynamically include state changes as part of the message
payload for all draw calls, rather than defining a specific new message for every possible
combination.

* Source/WebCore/platform/graphics/displaylists/DisplayListRecorder.cpp:
(WebCore::DisplayList::Recorder::strokePath):
* Source/WebCore/platform/graphics/displaylists/DisplayListRecorder.h:
* Source/WebCore/platform/graphics/displaylists/DisplayListRecorderImpl.cpp:
(WebCore::DisplayList::RecorderImpl::recordStrokeLineWithColorAndThickness):
* Source/WebCore/platform/graphics/displaylists/DisplayListRecorderImpl.h:
* Source/WebKit/GPUProcess/graphics/RemoteDisplayListRecorder.cpp:
(WebKit::RemoteDisplayListRecorder::strokeLineWithColorAndThickness):
* Source/WebKit/GPUProcess/graphics/RemoteDisplayListRecorder.h:
* Source/WebKit/GPUProcess/graphics/RemoteDisplayListRecorder.messages.in:
* Source/WebKit/Platform/IPC/StreamClientConnection.h:
* Source/WebKit/WebProcess/GPU/graphics/RemoteDisplayListRecorderProxy.cpp:
(WebKit::RemoteDisplayListRecorderProxy::recordStrokeLineWithColorAndThickness):
* Source/WebKit/WebProcess/GPU/graphics/RemoteDisplayListRecorderProxy.h:

Canonical link: <a href="https://commits.webkit.org/252522@main">https://commits.webkit.org/252522@main</a>
</pre>

### CODE REVIEW

File: Source/WebKit/Platform/IPC/StreamClientConnection.h [hunk 1]
```diff
@@ -157,7 +157,7 @@ class StreamClientConnection final {
         Semaphore clientWait;
     };
     std::optional<Semaphores> m_semaphores;
-    unsigned m_maxBatchSize { 40 }; // Number of messages marked as StreamBatched to accumulate before notifying the server.
+    unsigned m_maxBatchSize { 20 }; // Number of messages marked as StreamBatched to accumulate before notifying the server.
     unsigned m_batchSize { 0 };
```
Comment (weinig) [inline Source/WebKit/Platform/IPC/StreamClientConnection.h:161]: Was this intended as part of the change? The ChangeLog didn't mention it and it seems a bit unrelated.

Comment (mattwoodrow) [reply Source/WebKit/Platform/IPC/StreamClientConnection.h:161]: It is! You're right that the ChangeLog definitely should mention it though, I'll do that now.

**Review by smfr — APPROVED** [review]

File: Source/WebCore/platform/graphics/displaylists/DisplayListRecorder.cpp [hunk 2]
```diff
@@ -469,8 +469,15 @@ void Recorder::strokeRect(const FloatRect& rect, float lineWidth)
...
 {
-    appendStateChangeItemIfNecessary();
 #if ENABLE(INLINE_PATH_DATA)
+    auto& state = currentState().state;
+    if (state.changes() && state.containsOnlyInlineChanges() && !state.changes().contains(GraphicsContextState::Change::FillBrush) && path.hasInlineData() && path.hasInlineData<LineData>()) {
+        recordStrokeLineWithColorAndThickness(*strokeColor().tryGetAsSRGBABytes(), strokeThickness(), path.inlineData<LineData>());
+        return;
+    }
+
+    appendStateChangeItemIfNecessary();
```
Comment (shallawa) [inline Source/WebCore/platform/graphics/displaylists/DisplayListRecorder.cpp:479]: This call could be put before 

```
recordStrokeLine(path.inlineData<LineData>());
```

File: Source/WebCore/platform/graphics/displaylists/DisplayListRecorder.cpp [hunk 3]
```diff
@@ -482,6 +489,8 @@ void Recorder::strokePath(const Path& path)
             recordStrokeBezierCurve(path.inlineData<BezierCurveData>());
         return;
     }
+#else
+    appendStateChangeItemIfNecessary();
```
Comment (shallawa) [inline Source/WebCore/platform/graphics/displaylists/DisplayListRecorder.cpp:493]: And this one could be moved just before 

```
recordStrokePath(path);
```

So it is clearer that every time we record a drawing command, we are sure the state was already updated.
